drivers/ethernet/eth_gd32_hal.c

```diff
@@ -0,0 +1,842 @@
+{
+	int ret = 0;
+	int key;
+	ErrStatus hal_ret;
+
+	key = irq_lock();
+
+	enet_ptp_timestamp_update_config(ENET_PTP_ADD_TO_TIME, tm->second, tm->nanosecond);
+	hal_ret = enet_ptp_timestamp_function_config(ENET_PTP_SYSTIME_INIT);
+	if (hal_ret != SUCCESS) {
+		ret = -EIO;
+	}
+
+	irq_unlock(key);
+
+	return ret;
+}
+
+static int ptp_clock_gd32_get(const struct device *dev,
+			      struct net_ptp_time *tm)
+{
+	int key;
+	enet_ptp_systime_struct hal_tm, hal_tm_2;
+	uint32_t second_2;
+
+	key = irq_lock();
+
+	enet_ptp_system_time_get(&hal_tm);
+	tm->second = hal_tm.second;
+	tm->nanosecond = hal_tm.subsecond;
+
+	enet_ptp_system_time_get(&hal_tm_2);
+	second_2 = hal_tm_2.second;
+
+	irq_unlock(key);
+
+	if (tm->second != second_2 && tm->nanosecond < NSEC_PER_SEC / 2) {
+		/* Second rollover has happened during first measurement: second register
+		 * was read before second boundary and nanosecond register was read after.
+		 * We will use second_2 as a new second value.
+		 */
+		tm->second = second_2;
+	}
+
+	return 0;
+}
+
+static int ptp_clock_gd32_adjust(const struct device *dev, int increment)
+{
+	int ret = 0;
+	int key;
+	ErrStatus hal_ret;
+
+	if ((increment <= (int32_t)(-NSEC_PER_SEC)) ||
+			(increment >= (int32_t)NSEC_PER_SEC)) {
+		ret = -EINVAL;
+	} else {
+		key = irq_lock();
+
+		if (increment >= 0) {
+			enet_ptp_timestamp_update_config(ENET_PTP_ADD_TO_TIME, 0, increment);
+		} else {
+			enet_ptp_timestamp_update_config(ENET_PTP_SUBSTRACT_FROM_TIME, 0,
+							 increment);
+		}
+		hal_ret = enet_ptp_timestamp_function_config(ENET_PTP_SYSTIME_UPDATE);
+		if (hal_ret != SUCCESS) {
+			ret = -EIO;
+		}
+
+		irq_unlock(key);
+	}
+
+	return ret;
+}
+
+static int ptp_clock_gd32_rate_adjust(const struct device *dev, double ratio)
+{
+	struct ptp_context *ptp_context = dev->data;
+	struct eth_gd32_hal_dev_data *eth_dev_data = ptp_context->eth_dev_data;
+	int key, ret;
+	uint32_t addend_val;
+	ErrStatus hal_ret;
+
+	/* No change needed */
+	if (ratio == 1.0f) {
+		return 0;
+	}
+
+	key = irq_lock();
+
+	ratio *= eth_dev_data->clk_ratio_adj;
+
+	/* Limit possible ratio */
+	if (ratio * 100 < CONFIG_ETH_GD32_HAL_PTP_CLOCK_ADJ_MIN_PCT ||
+			ratio * 100 > CONFIG_ETH_GD32_HAL_PTP_CLOCK_ADJ_MAX_PCT) {
+		ret = -EINVAL;
+		goto error;
+	}
+
+	/* Save new ratio */
+	eth_dev_data->clk_ratio_adj = ratio;
+
+	/* Update addend register */
+	addend_val = UINT32_MAX * eth_dev_data->clk_ratio * ratio;
+
+	enet_ptp_timestamp_addend_config(addend_val);
+	hal_ret = enet_ptp_timestamp_function_config(ENET_PTP_ADDEND_UPDATE);
+	if (hal_ret != SUCCESS) {
+		ret = -EIO;
+		goto error;
+	}
+
+	ret = 0;
+
+error:
+	irq_unlock(key);
+
+	return ret;
+}
+
+static const struct ptp_clock_driver_api api = {
+	.set = ptp_clock_gd32_set,
+	.get = ptp_clock_gd32_get,
+	.adjust = ptp_clock_gd32_adjust,
+	.rate_adjust = ptp_clock_gd32_rate_adjust,
+};
+
+static int ptp_gd32_init(const struct device *port)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(mac));
```

Comment:
As you are getting the ethernet device here, why use `DT_NODELABEL()`, seems like this could be `DEVICE_DT_INST_GET`

Reply:
Done